aptos-move/framework/aptos-framework/sources/transaction_validation.move

```diff
@@ -70,6 +72,25 @@ module aptos_framework::transaction_validation {
         });
     }
 
+    /// Called in prologue to optionally hold some amount for special txns (e.g. randomness txns).
+    /// `release_to_balance()` should be invoked in the corresponding epilogue with the same arguments.
+    fun collect_deposit(gas_payer: address, amount: Option<u64>) {
+        if (option::is_some(&amount)) {
+            let amount = option::extract(&mut amount);
+            let balance = coin::balance<AptosCoin>(gas_payer);
+            assert!(balance >= amount, error::invalid_state(PROLOGUE_EINSUFFICIENT_BALANCE_FOR_REQUIRED_DEPOSIT));
```

Comment:
this will need to be changed based on https://github.com/aptos-labs/aptos-core/pull/12745 to use is_balance_at_least()